internal/worker/worker.go

```diff
@@ -0,0 +1,444 @@
+			if w.concurrencyLimit() {
+				continue
+			}
+
+			w.processEvents(ctx)
+
+		case <-ctx.Done():
+			if w.dispatched > 0 {
+				continue
+			}
+
+			break Loop
+		}
+	}
+}
+
+func (w *Worker) processEvents(ctx context.Context) {
+	// XXX: consider having a separate context for message retrieval
+	msgs, err := w.stream.PullMsg(ctx, 1)
+
+	switch {
+	case err == nil:
+	case errors.Is(err, nats.ErrTimeout):
+		w.logger.WithFields(
+			logrus.Fields{"err": err.Error()},
+		).Trace("no new events")
+	default:
+		w.logger.WithFields(
+			logrus.Fields{"err": err.Error()},
+		).Warn("retrieving new messages")
+
+		metrics.NATSError("pull-msg")
+	}
+
+	for _, msg := range msgs {
+		if ctx.Err() != nil || w.concurrencyLimit() {
+			w.eventNak(msg)
+
+			return
+		}
+
+		// spawn msg process handler
+		w.syncWG.Add(1)
+
+		go func(msg events.Message) {
+			defer w.syncWG.Done()
+
+			atomic.AddInt32(&w.dispatched, 1)
+			defer atomic.AddInt32(&w.dispatched, -1)
+
+			w.processSingleEvent(ctx, msg)
+		}(msg)
+	}
+}
+
+func (w *Worker) processSingleEvent(ctx context.Context, e events.Message) {
+	// extract parent trace context from the event if any.
+	ctx = e.ExtractOtelTraceContext(ctx)
+
+	ctx, span := otel.Tracer(pkgName).Start(
+		ctx,
+		"worker.processSingleEvent",
+	)
+	defer span.End()
+
+	condition, err := conditionFromEvent(e)
+	if err != nil {
+		w.logger.WithError(err).WithField("subject", e.Subject()).Warn("unable to retrieve condition from message")
+
+		metrics.RegisterEventCounter(false, "ack")
+		w.eventAckComplete(e)
+
+		return
+	}
+
+	// check and see if the task is or has-been handled by another worker
+	currentState, err := rkv.ConditionStatus(
+		condition.ID.String(),
+		w.facilityCode,
+		inventoryStatusKVBucket,
+		events.AsNatsJetStreamContext(w.stream.(*events.NatsJetstream)),
+	)
+
+	// errors from ConditionStatus are for logging purposes only
+	if err != nil {
+		w.logger.WithField("conditionID", condition.ID.String()).Warn(err)
+	}
+
+	switch currentState {
+	case rkv.InProgress:
+		w.logger.WithField("conditionID", condition.ID.String()).Info("condition is already in progress")
+		w.eventAckInProgress(e)
+		metrics.RegisterSpanEvent(span, condition, w.id.String(), "", "ackInProgress")
+
+		return
+
+	case rkv.Complete:
+		w.logger.WithField("conditionID", condition.ID.String()).Info("condition is complete")
+		w.eventAckComplete(e)
+		metrics.RegisterSpanEvent(span, condition, w.id.String(), "", "ackComplete")
+
+		return
+
+	case rkv.Orphaned:
+		w.logger.WithField("conditionID", condition.ID.String()).Warn("restarting this condition")
+		metrics.RegisterSpanEvent(span, condition, w.id.String(), "", "restarting condition")
+
+	// we need to restart this event
+	case rkv.NotStarted:
+		w.logger.WithField("conditionID", condition.ID.String()).Info("starting new condition")
+		metrics.RegisterSpanEvent(span, condition, w.id.String(), "", "start new condition")
+
+	// break out here, this is a new event
```

Comment:
This comment should be moved up, as is it feels like it applies to `Indeterminate` and not `NotStarted`. Same thing for `Orphaned`.